src/Nethermind/Nethermind.Evm/TransactionProcessing/TransactionProcessor.cs

```diff
@@ -360,21 +383,25 @@ private void Execute(Transaction transaction, BlockHeader block, ITxTracer txTra
                     {
                         _stateProvider.DecrementNonce(caller);
                     }
+
+                    _stateProvider.Commit(spec);
                 }
-                
-                _stateProvider.Commit(spec);
+            }
+            else if (commit)
+            {
+                _storageProvider.Commit(txTracer.IsTracingState ? txTracer : NullStorageTracer.Instance);
+                _stateProvider.Commit(spec, txTracer.IsTracingState ? txTracer : NullStateTracer.Instance);
             }
 
-            if (!isCall && notSystemTransaction)
+            if (!restore && notSystemTransaction)
```

Comment:
how about commit?

Reply:
either if we commit or not we want to build state